[seantehds] iP

seantehds opened the pull request:
# Emiya Taskbot

> "People die when they are killed!" - Shirou Emiya [(source)](https://knowyourmeme.com/memes/people-die-if-they-are-killed)

The Emiya Taskbot is:
* text-based
* easy to learn
* ~~able to cook for you~~ __fast to use__

All that needs to be done is,

1. download it from [here](https://github.com/seantehds/ip/releases/tag/v0.1).
2. double-click it
3. add your tasks.
4. let it manage your tasks for you 😄 

And it is **FREE** (for now 😈)!

Features:
- [X] Managing tasks
- [ ] Manage deadlines (coming soon) (maybe)
- [ ] Reminders (coming soon) (also maybe)

RANDOM JAVA CODE GO:
Random `main` method:
```
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

seantehds committed:
Added basic functionality to print output for new chatbot Emiya

seantehds committed:
Added in basic input recognition from user and displaying their input as an output on the screen

seantehds committed:
Added in list functionality; Also able to detect if list is empty and ask user to add items

seantehds committed:
Added in mark and unmark functionality to the list, as well as the Task class for individual tasks

seantehds committed:
Added in new subclasses for Task for the 3 different event types, as well as updating the app to handle different Task types

seantehds committed:
added basic tests to test functionality using runtestemiya.bat file

seantehds committed:
Added exception handling to the Emiya chatbot, such as from unknown commands or wrong formats for commands

seantehds committed:
Changed how tasks were represented inside Emiya, as well as added delete functionality

seantehds committed:
Fix a few bugs within Emiya with different numbers of tasks, as well as identation levels

seantehds committed:
Refactor EmiyaExceptions to be in a separate package

seantehds committed:
Change import statements and refactor Task into the task package

seantehds committed:
Add ability to retrieve the absolute path name to root directory

seantehds committed:
Add checking of existence of data folder within root directory

seantehds committed:
Add ability to create the data directory if it does not exist

seantehds committed:
Add functionality to write file to data directory if it does not exist

seantehds committed:
Add ability to write tasks into emiya.txt after commands by user

seantehds committed:
Add functionality to read from the saved file to populate tasks

seantehds committed:
Fix bugs when writing to saved emiya.txt

seantehds committed:
Add the DataHandler class to handle date inputs for deadlines and events

seantehds committed:
Add error handling for user inputs for dates

seantehds committed:
Merge branch 'branch-Level-7'

seantehds committed:
Merge branch 'branch-Level-8'

seantehds committed:
Add TaskList, Ui class for more OOP

seantehds committed:
Add Parser class to handle parsing of inputs

seantehds committed:
Change handling of parsing of text from DateHandler to Parser

seantehds committed:
Clean comments and unnecessary print statements

seantehds committed:
Merge remote-tracking branch 'origin/add-gradle-support'

seantehds committed:
Set up Gradle by merging with add-gradle branch

seantehds committed:
Add unit tests for Parser and DateHandler

seantehds committed:
Fix bug when printing output when adding task to list

seantehds committed:
Add JavaDoc for DateHandler class

seantehds committed:
Fix issue where certain keywords would not cause specific exceptions to be thrown

seantehds committed:
Add JavaDoc to emiyaexception, logic, parser and storage packages

seantehds committed:
Change code to fit coding standards specified

seantehds committed:
Merge branch 'branch-A-JavaDoc'

seantehds committed:
Merge branch 'branch-A-CodingStandard'

ldinghan commented on src/main/java/emiya/emiyaexception/EmptyDeleteException.java:
```diff
@@ -0,0 +1,9 @@
+package emiya.emiyaexception;
+
+public class EmptyDeleteException extends EmiyaException{
```
Would be good to add javadocs to your exception classes too! There are quite a few of exception classes without javadocs headers

ldinghan commented on src/main/java/emiya/parser/Parser.java:
```diff
@@ -0,0 +1,145 @@
+ * A class that takes in a String input from the user and parses it to determine what command
+ * the user has given.
+ */
+public class Parser {
```
Parser is really well done! Looks clean and readable

ldinghan commented on src/main/java/emiya/storage/Storage.java:
```diff
@@ -0,0 +1,165 @@
+     * @param dirName The name of the directory that contains the file to be accessed.
+     * @return A String containing the contents of the file indicated by fileName.
+     */
+    public String fileContents(String fileName, String dirName) {
```
Perhaps another method name as a verb would be more suitable?

ldinghan commented on src/main/java/emiya/task/Deadline.java:
```diff
@@ -0,0 +1,39 @@
+    }
+
+    @Override
+    public String toString() {
```
Maybe instead of having an if-else that returns almost the same thing, it might be shorter if you could find a way to place either a " "(space) or a "X" within the brackets depending on the boolean isCompleted and only having one return statement!

ldinghan reviewed (commented):
Looks really good overall! The classes were nicely split and follows coding standard very closely! More javadocs could definitely be added though, especially with the exception classes, Ui class, TaskList class, etc.

seantehds committed:
Add a GUI to Emiya using JavaFX

seantehds committed:
Merge branch 'branch-Level-10'

adammangzijun commented on src/main/java/emiya/commands/DeadlineCommand.java:
```diff
@@ -0,0 +1,53 @@
+ */
+public class DeadlineCommand {
+
+    /**
```
Is there a way to reduce the number of parameters accepted? It makes the code too lengthy, especially the javaDoc.

adammangzijun commented on src/main/java/emiya/commands/EventCommand.java:
```diff
@@ -0,0 +1,55 @@
+ */
+public class EventCommand {
+
+    /**
```
Every Command is made such that you accept too many arguments, as such your javaDoc becomes very chunky.

adammangzijun commented on src/main/java/emiya/commands/EventCommand.java:
```diff
@@ -0,0 +1,55 @@
+        taskList.add(event);
+        storage.writeToFileFromTaskList(taskList, fileName, dirName);
+        if (taskList.size() == 1) {
+            return ui.addedSingularMessage(event, taskList);
```
Methods should be named as a verb, in this case, it is better to put it as addSingularMessage.

adammangzijun commented on src/main/java/emiya/commands/EventCommand.java:
```diff
@@ -0,0 +1,55 @@
+        if (taskList.size() == 1) {
+            return ui.addedSingularMessage(event, taskList);
+        } else {
+            return ui.addedPluralMessage(event, taskList);
```
Same for this.

adammangzijun commented on src/main/java/emiya/commands/MarkCommand.java:
```diff
@@ -0,0 +1,42 @@
+            }
+            taskList.get(pos - 1).setMarked();
+            storage.writeToFileFromTaskList(taskList, fileName, dirName);
+            return ui.markedMessage(pos, taskList);
```
markedMessage -> markMessage

adammangzijun commented on src/main/java/emiya/storage/Storage.java:
```diff
@@ -0,0 +1,164 @@
+        File dataDir = new File(pathToDataDir);
+
+        if (!dataDir.exists()) {
+            boolean result = dataDir.mkdirs();
```
Instead of using general terms like result, use has, is or was for boolean variables.

adammangzijun commented on src/main/java/emiya/emiyaexception/CreateDirectoryFailException.java:
```diff
@@ -0,0 +1,12 @@
+ * An exception that is thrown when the program is unable to create a new directory.
+ * This is used to create the directory that the list contents will be saved into.
+ */
+public class CreateDirectoryFailException extends EmiyaException {
```
Class names should be nouns.

adammangzijun commented on src/main/java/emiya/parser/Parser.java:
```diff
@@ -0,0 +1,146 @@
+            throw new InvalidDateException();
+        }
+
+        return new String[] {datePart, hour < 10 ? "0" + hour : hour + "", min < 10 ? "0" + min : min + ""};
```
this 2nd parameter can be separated into another variable for easy readability.

adammangzijun reviewed (commented):
Overall well done in code quality, just minor issues to take note of!! 😄

seantehds committed:
Improve code quality

There are areas within the code that can be of higher quality.

Improving code quality is important in making the code more readable
and easier to understand for anyone who reads the code.

Let's improve the quality of the code in order to be able to achieve
that goal,
* by minimizing scope of variables as much as possible
* by making happy paths more prominent whenever possible

Doing this is critical to improving the quality of the code.

seantehds committed:
Add assertions into key parts of the code

The code does not have any assertions statements within at this moment.

Assertions in Java are useful in order to determine whether the code
is working as intended or not.

Let's add in assertions into the existing code in order to ensure that
the code is working as intended at different points within the code.

The assertion statements are being added at key points to ensure that
the application is working as intended and can carry out what the task
it needs to do.

seantehds committed:
Merge pull request #1 from seantehds/branch-A-Assertions

Add assertions into key parts of the code

seantehds committed:
Pull from remote repo to fix merge conflicts

seantehds committed:
Merge branch 'branch-A-CodeQuality' into master

seantehds committed:
Add help functionality to the application

When a user opens the application, it does not provide any prompts or
guide on how the application can be used. As such, the application
may seem intimidating and unfriendly to new users, or users who
have forgotten the commands.

Let's add,
* add a welcome message to make the user feel more welcome, as
well as introduce the help function to the user
* add a help function to the application to allow the user
to see all commands that the application has to offer

Doing this is important in order to allow new users to learn how to
use the application, while providing existing users a refresher
in case they have forgotten the commands needed to use the application.

seantehds committed:
Merge branch 'branch-C-Help'

seantehds committed:
Fix bug when printing help command

seantehds committed:
Fix bug when displaying deadline tasks

seantehds committed:
Change application background colour and add product screenshot

seantehds committed:
Add user guide documentation for the application

seantehds committed:
Remove additional columns from the markdown table in the user guide

seantehds committed:
Fix bug when displaying dates, add new exceptions for mark and unmark and remove old text-based tests

seantehds committed:
Add bye command as well as fixed bug with program not recognising /by, /from and /to in user inputs

seantehds committed:
Add more unit tests for parser methods and change name of InvalidDateException

seantehds committed:
Update user guide to contain bye command

seantehds committed:
Update files for checkstyle

seantehds committed:
Fix bug when user uses invalid date/time
